bundles/org.openhab.binding.heos/src/main/java/org/openhab/binding/heos/internal/handler/HeosBridgeHandler.java

```diff
@@ -0,0 +1,466 @@
+        logger.debug("Logging in to HEOS account.");
+        HeosResponseObject<Void> response = connection.logIn(username, password);
+
+        if (response.result) {
+            logger.debug("successfully logged-in, event is fired to handle post-login behaviour");
+            return;
+        }
+
+        HeosError error = response.getError();
+        logger.debug("Failed to login: {}", error);
+        updateStatus(ONLINE, ThingStatusDetail.CONFIGURATION_ERROR,
+                error != null ? error.code.toString() : "Failed to login, no error was returned.");
+    }
+
+    @Override
+    public void dispose() {
+        bridgeHandlerDisposalOngoing = true; // Flag to prevent the handler from being updated during disposal
+
+        terminateStartupSequence();
+
+        HeosFacade localApiConnection = apiConnection;
+        if (localApiConnection == null) {
+            logger.debug("Not disposing bridge because of missing apiConnection");
+            return;
+        }
+
+        localApiConnection.unregisterForChangeEvents(this);
+        logger.debug("HEOS bridge removed from change notifications");
+
+        logger.debug("Dispose bridge '{}'", thing.getProperties().get(PROP_NAME));
+        localApiConnection.closeConnection();
+    }
+
+    private void terminateStartupSequence() {
+        ScheduledFuture<?> localStartupFuture = startupFuture;
+        if (localStartupFuture != null && !localStartupFuture.isCancelled()) {
+            localStartupFuture.cancel(true);
+        }
+    }
+
+    /**
+     * Manages the removal of the player or group channels from the bridge.
+     */
+    @Override
+    public synchronized void childHandlerDisposed(ThingHandler childHandler, Thing childThing) {
+        logger.debug("Disposing child handler for: {}.", childThing.getUID().getId());
+        if (bridgeHandlerDisposalOngoing) { // Checks if bridgeHandler is going to disposed (by stopping the binding or
+            // openHAB for example) and prevents it from being updated which stops the
+            // disposal process.
+        } else if (childHandler instanceof HeosPlayerHandler) {
+            String channelIdentifier = "P" + childThing.getUID().getId();
+            updateThingChannels(channelManager.removeSingleChannel(channelIdentifier));
+        } else if (childHandler instanceof HeosGroupHandler) {
+            String channelIdentifier = "G" + childThing.getUID().getId();
+            updateThingChannels(channelManager.removeSingleChannel(channelIdentifier));
+            // removes the handler from the groupMemberMap that handler is no longer called
+            // if group is getting online
+            removeGroupHandlerInformation((HeosGroupHandler) childHandler);
+        }
+    }
+
+    @Override
+    public void childHandlerInitialized(ThingHandler childHandler, Thing childThing) {
+        scheduler.schedule(() -> addPlayerChannel(childThing, null), 5, TimeUnit.SECONDS);
```

Comment:
Are you referring to the async nature or the actual delay? Do you have other suggestions of maybe additional safeguards we could put in place? 

The framework requires us to return quickly from handlers such as this one and the HEOS API is not quick to respond quickly without doing it async.

Reply:
My question is why a lot of your async tasks are scheduled to occur after several seconds of delay?

Reply:
Some are because of error handling but indeed I spotted at least 5 for which I do (today) not see a reason why they should not be submitted right away. I will take a look at it I hope before the end of this week :-)